### PR TITLE
Add tenant-scoped filestore APIs and workspace storage management

### PR DESCRIPTION
**Pull request by arthur-zhang**
## Summary

- Add filestore APIs with token-based authentication, tenant isolation, filesystem provisioning, scanning, cleanup, and cursor support
- Add workspace storage usage tracking and migrations
- Integrate filestore authentication and storage handling across files, sessions, batches, memory, skills, and deployment flows
- Seed built-in skills and update backend design documentation

## Testing

- Added and updated unit, database, integration, and API coverage for filestore behavior, authentication, storage usage, migrations, and affected resources
- Not run (not requested)

## Summary by CodeRabbit

* **New Features**
  * Added Filestore support with directory and file operations, streaming uploads, ranged reads, copying, moving, deletion, pagination, quotas, and background cleanup.
  * Added secure read-write and read-only Filestore tokens with dedicated authentication and protocol errors.
  * Added a CLI for issuing Filestore tokens.
  * Added multi-bucket object storage support, version-aware deletion, and richer upload/download metadata.

* **Bug Fixes**
  * Improved storage accounting, transactional file updates, cleanup retries, and orphaned-object handling.

* **Documentation**
  * Updated authentication, runtime configuration, Filestore, database, and migration guidance.

* **Tests**
  * Expanded coverage for authentication, storage, Filestore workflows, quotas, migrations, and integration scenarios.

### CODE REVIEW

**Comment by jh0904** [comment]
@codex review

**Comment by arthur-zhang** [comment]
@duckpr

**Comment by arthur-zhang** [comment]
@duckpr

**Comment by arthur-zhang** [comment]
@duckpr

**Comment by arthur-zhang** [comment]
@duckpr

File: internal/filestore/service.go [hunk 1]
```diff
@@ -0,0 +1,772 @@
...
+		return readFileResult{Body: io.NopCloser(bytes.NewReader(nil)), MediaType: mediaType}, nil
+	}
+	object, err := s.store.Open(ctx, *entry.S3Key, objectRange)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return readFileResult{}, internalError("read file object", errors.New("object metadata exists but blob is missing"))
+		}
+		return readFileResult{}, mapBlobstoreError("read file", err)
+	}
+	return readFileResult{Body: object.Body, Size: object.Size, MediaType: mediaType}, nil
```
Comment (arthur-zhang) [inline internal/filestore/service.go]: **[bug]** `ReadFile` returns `Size: object.Size` from S3. `storage.Object` documents `Size == -1` when the provider omits `Content-Length` (also called out in the runtime-configuration design updates in this PR). The handler then does `Content-Length: -1` and `io.CopyN(w, body, result.Size)`. Go's `CopyN` rejects negative `n`, so a valid object with unknown length becomes a truncated/broken download after headers are committed. The service already knows the authoritative size from `entry.SizeBytes` and the resolved range, but discards both.

**Suggestion:** Have `resolveReadRange` return the exact byte count to stream (or compute it beside the range). Prefer DB/range size over S3 `Content-Length`; only fall back to S3 when metadata is absent. In the handler, omit `Content-Length` and use `io.Copy` when size is unknown—matching the platform preview behavior already documented for storage.

File: internal/filestore/handler.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+	}
+	if result.Body != nil {
+		defer result.Body.Close()
+	}
+	if result.MediaType != "" {
+		w.Header().Set("Content-Type", result.MediaType)
+	} else {
+		w.Header().Set("Content-Type", "application/octet-stream")
+	}
+	w.Header().Set("Content-Length", strconv.FormatInt(result.Size, 10))
```
Comment (arthur-zhang) [inline internal/filestore/handler.go]: **[bug]** Companion to Issue 1: `readFile` always sets `Content-Length` from `result.Size` and always uses `io.CopyN`. Even when service-side size is fixed, this handler assumes every successful open has a non-negative, exact length. That is a brittle HTTP boundary for a streaming resource and contradicts the storage layer’s `-1` contract.

**Suggestion:** Stream with known length only when `result.Size >= 0`; otherwise stream without `Content-Length`. Add a handler/service unit test where Open returns `Size: -1` and a non-empty body.

File: internal/filestore/service.go [hunk 3]
```diff
@@ -0,0 +1,772 @@
...
+		return db.FilestoreFilesystem{}, permissionDenied("Filestore token does not grant access to this filesystem")
+	}
+	return filesystem, nil
+}
+
+func (s *Service) requireParentDirectory(ctx context.Context, workspaceID, filesystemID int64, entryPath string) *apiError {
+	parent := parentPath(entryPath)
+	entry, err := s.db.GetFilestoreEntry(ctx, workspaceID, filesystemID, parent)
+	if errors.Is(err, db.ErrNotFound) {
+		return failedPrecondition("parent directory does not exist")
```
Comment (arthur-zhang) [inline internal/filestore/service.go:473]: **[bug]** Missing-parent failures are mapped inconsistently. `requireParentDirectory` (CreateFile/CopyFile pre-check) returns `failed_precondition`, but the authoritative in-transaction check `requireFilestoreDirectoryTx` surfaces `db.ErrNotFound`, which `mapDatabaseError` turns into `not_found`. Clients therefore see different codes for the same semantic failure depending on whether the parent vanished before upload, during upload, or only at commit—and MakeDirectory/Put paths always use `not_found`.

**Suggestion:** Map “parent path missing or not a directory” to one stable code at the service/DB boundary (prefer `failed_precondition` for mkdir-parent semantics, and use a dedicated `db` error rather than reusing generic `ErrNotFound`). Drop the pre-check’s divergent mapping, or make Put/Copy return the same error type as the pre-check.

File: internal/filestore/service.go [hunk 4]
```diff
@@ -0,0 +1,772 @@
...
+	})
+	if errors.Is(err, db.ErrNotFound) {
+		return nil
+	}
+	return mapDatabaseErrorOrNil("remove directory", err)
+}
+
+// CreateFile 流式上传文件并写入元数据。上传前先登记延迟清理哨兵，
+// 使进程在“对象已写入、数据库尚未提交”之间崩溃时仍能回收孤儿对象。
+func (s *Service) CreateFile(ctx context.Context, principal Principal, params createFileParams, body io.Reader) (fileResponse, *apiError) {
```
Comment (arthur-zhang) [inline internal/filestore/service.go:174]: **[suggestion]** `CreateFile` and `CopyFile` duplicate the same multi-step orchestration: resolve FS → preflight parent → mint unique key → enqueue orphan guard → write object → attach version (with discard on failure) → DB bind → encode payload. That is the most failure-sensitive path in the package; two hand-rolled copies invite drift (e.g. one side forgetting attach-before-put, discard options, or version binding).

**Suggestion:** Extract a single helper, e.g. `commitGuardedBlob(ctx, principal, filesystem, key, write func() (etag, versionID, size, error), put func(blob) (...))`, used by both create and copy. Keep only the source-read / hashing differences outside the helper. This is high-value code judo: one lifecycle, two call sites.

File: internal/filestore/cleanup.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+	FailLeasedFilestoreObjectCleanupJob(context.Context, int64, string, string, time.Duration, int) error
+	ExpireFilestoreEntries(context.Context, int) ([]db.FilestoreObjectCleanupJob, error)
+}
+
+// StartFilestoreCleanupWorker 启动单个后台循环，分别处理对象清理任务与 TTL 到期扫描。
+// 两类工作都会在启动时立即执行一次，不必等待首个定时周期。
+func StartFilestoreCleanupWorker(
+	ctx context.Context,
+	database filestoreCleanupDatabase,
+	store storage.ObjectStore,
```
Comment (arthur-zhang) [inline internal/filestore/cleanup.go]: **[suggestion]** The PR introduces multi-bucket `storage.Client.ForBucket` and the generic object-cleanup worker correctly selects a store per job bucket (`internal/cleanup/worker.go`). Filestore cleanup deliberately takes a single `ObjectStore` and permanently parks any job whose `job.Bucket != store.Name()` with a hard failure delay. That contradicts the design doc’s multi-bucket cleanup story and the main.go comment that “通用清理任务可通过 client 按任务记录选择其他 bucket,” while Filestore is left on the weaker path. Today everything shares the default bucket, so this is latent, but the abstraction is already wrong for the capability the PR advertises.

**Suggestion:** Inject `storage.Client` (like the generic cleanup worker), call `ForBucket(job.Bucket)` per job, and delete the bucket-mismatch special case. Drop the unused `cfg config.Config` parameter while you are there—it is threaded through the whole cleanup loop and never read.

File: internal/db/filestore_scan.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+		return FilestoreFilesystem{}, err
+	}
+	return databaseRow.filesystem(), nil
+}
+
+func filestoreEntrySelectSQL() string {
+	return "select " + filestoreEntryColumns() + " from filestore_entries"
+}
+
+func filestoreEntryColumns() string {
```
Comment (arthur-zhang) [inline internal/db/filestore_scan.go:57]: **[suggestion]** Entry/filesystem column lists and scanning exist twice: positional PGX (`filestore_scan.go`) and named sqlx (`filestore_sqlx.go`). Migrations 00018–00024 already rewrote Filestore schema multiple times (bigint → UUID refs, column reorder, session uniqueness). Each schema change must be hand-synced across two SELECT strings, two scanners, and row DTOs—high drift risk for a brand-new surface.

**Suggestion:** Keep one canonical column projection (shared string or shared row struct) and adapt only the execution style (pgx `QueryRow` vs sqlx named). At minimum, generate both SELECTs from a single ordered field list so cast/alias changes cannot diverge.

File: internal/db/filestore_entries.go [hunk 7]
```diff
@@ -0,0 +1,897 @@
...
+	}
+	if err := tx.Commit(ctx); err != nil {
+		return FilestoreEntry{}, err
+	}
+	return directory, nil
+}
+
+// PutFilestoreFile 将对象元数据写入命名空间，并与配额核算、旧对象清理任务及
+// 孤儿哨兵取消保持同一事务边界。
+func (d *DB) PutFilestoreFile(ctx context.Context, input PutFilestoreFileInput) (FilestoreMutationResult, error) {
```
Comment (arthur-zhang) [inline internal/db/filestore_entries.go]: **[suggestion]** `filestore_entries.go` is ~897 lines of mutation logic (mkdir/put/copy/move/remove + locks + path helpers). It is one modest feature away from crossing the 1k-line hard stop. The public methods share a repeated open-coded pattern: validate → `beginFilestoreNamespaceMutation` → work → commit, with large inline SQL.

**Suggestion:** Split by responsibility before it grows: e.g. `filestore_entries_read.go`, `filestore_entries_write.go` (put/copy), `filestore_entries_namespace.go` (move/remove), and keep lock/begin helpers shared. Extract the repeated “begin + defer rollback + commit” wrapper so each mutation is the body only.

File: internal/filestore/service_test.go [hunk 8]
```diff
@@ -0,0 +1,1184 @@
+package filestore
```
Comment (arthur-zhang) [inline internal/filestore/service_test.go:1]: **[suggestion]** New production-adjacent test file is already 1184 lines (fakes, path rejection, token binding, create/copy/move/read, orphan cleanup, payload encoding). That violates the “do not push a file over 1k without a strong reason” bar and makes review/navigation of the most important service regressions harder than necessary.

**Suggestion:** Split along natural seams: `service_validation_test.go`, `service_write_test.go`, `service_read_test.go`, `service_fakes_test.go` (shared fakes). Keep table-driven cases; just remove the monolithic file.

File: internal/filestore/service.go [hunk 9]
```diff
@@ -0,0 +1,772 @@
...
+			S3Bucket:              s.store.Name(),
+			S3Key:                 objectKey,
+			S3ETag:                upload.ETag,
+			S3VersionID:           upload.VersionID,
+			ExpiresAt:             expiresAt,
+		},
+		OverwriteExisting:          params.OverwriteExisting,
+		OrphanCleanupJobExternalID: cleanupJob.ExternalID,
+		WorkspaceStorageLimitBytes: s.cfg.Storage.WorkspaceLimitBytes,
+		Actor:                      db.FilestoreActor{},
```
Comment (arthur-zhang) [inline internal/filestore/service.go]: **[suggestion]** Every mutation passes `Actor: db.FilestoreActor{}`. The DB layer carefully resolves `created_by_api_key_uuid` / session / code_session from actor IDs, but Filestore JWT principals never populate those fields (and `Principal` does not even carry API-key/session IDs—only account/filesystem scope). The result is permanent NULL audit columns on all Filestore entries, while a full actor plumbing stack remains in write paths.

**Suggestion:** Either (a) map principal subject/account/session into actor fields for real provenance, or (b) delete `FilestoreActor` from Filestore write inputs until a real caller exists. Half-wired audit plumbing is worse than none: it suggests coverage that is not there.

File: internal/filestore/service.go [hunk 10]
```diff
@@ -0,0 +1,772 @@
...
+	if authorization != nil {
+		authorizationJSON, err = json.Marshal(authorization)
+		if err != nil {
+			return nil, nil, invalidArgument("authorizationMetadata is not valid JSON")
+		}
+	}
+	return metadataJSON, authorizationJSON, nil
+}
+
+func fileDownloadable(metadata *authorizationMetadata) bool {
```
Comment (arthur-zhang) [inline internal/filestore/service.go:625]: **[suggestion]** `fileDownloadable` returns `true` when `authorizationMetadata` is omitted, but `false` when the object is present without an explicit `downloadable` field (`bool` zero value + `omitempty`). Clients that send `{"intent":"..."}` only will silently mark files non-downloadable; clients that omit the whole object get downloadable files. That is a sharp edge in the wire contract with no test calling it out.

**Suggestion:** Use `*bool` for `Downloadable` so absent means default-true, or default-true whenever the field is missing after decode. Add a unit test for `{intent: x}` without `downloadable`.

File: internal/filestore/service.go [hunk 11]
```diff
@@ -0,0 +1,772 @@
...
+		}
+	}
+	payload := filePayload{
+		UUID:              entry.UUID,
+		CreatedAt:         formatTimestamp(entry.CreatedAt),
+		Size:              protoInt64(*entry.SizeBytes),
+		MediaType:         stringValue(entry.MediaType),
+		Metadata:          metadata,
+		MD5:               stringValue(entry.MD5),
+		WorkspaceTaggedID: entry.ExternalID,
```
Comment (arthur-zhang) [inline internal/filestore/service.go]: **[suggestion]** Wire field `workspaceTaggedId` is filled from `entry.ExternalID` (`fse_...`), while the same name in JWT claims means `workspace.external_id`. There is no separate response field for entry external ID, and no test asserts the value. If rclone expects the workspace tagged ID, this is a contract bug; if it expects the entry tagged ID, the field name collides with auth terminology and will confuse every future reader.

**Suggestion:** Confirm against the rclone-filestore proto/fixture. If workspace ID is required, pass `principal.WorkspaceExternalID` (or filesystem’s workspace) into payload encoding. If entry external ID is required, rename the Go field/comment to `EntryTaggedID` / document the wire alias explicitly, and add a golden response test.

File: internal/db/migrations/00018_add_filestore.sql [hunk 12]
```diff
@@ -0,0 +1,145 @@
+-- +goose Up
```
Comment (arthur-zhang) [inline internal/db/migrations/00018_add_filestore.sql:1]: **[suggestion]** The feature ships as seven sequential migrations (00018–00024) that create bigint-scoped tables, then repeatedly rewrite them to UUID refs, reorder columns, and rebuild uniqueness. For a not-yet-production surface this is evolutionary residue that burdens every fresh install, review, and failure diagnosis (especially ACCESS EXCLUSIVE rebuilds and backfills in 00023/00024).

**Suggestion:** If these have not been applied to shared environments, squash into one or two final-shape migrations (schema + session backfill/invariant). If they must remain, add a short design note that intermediate shapes are historical only and must not be reintroduced in application code.

File: internal/filestore/service_types.go [hunk 13]
```diff
@@ -0,0 +1,9 @@
+package filestore
```
Comment (arthur-zhang) [inline internal/filestore/service_types.go]: **[nit]** `service_types.go` only holds `readFileResult`, while request/response DTOs live in `protocol.go` and the rest of service helpers live in `service.go`. `internal/filestore/blobstore/` is an empty package directory. Both are vestigial structure that look like abandoned extractions.

**Suggestion:** Move `readFileResult` next to `ReadFile`/`protocol` and delete the empty `blobstore` directory (or actually place a type there if a future adapter is planned—don’t leave empty trees).
